Anisotropic surfaces + Separated Infills (remake)

pi-squared-studio opened the pull request:
# Description

This package is a recompilation of this PR https://github.com/OrcaSlicer/OrcaSlicer/pull/10314 and it contains only a few implemented common functions: Anisotropic surfaces, Center of surface patterns and Separated Infills.
Also, all functions were individually tested for compatibility and various bugs were identified. 

## Anisotropic surfaces
This addition is used to apply unidirectional patterns to the upper and lower surfaces. Often, multidirectional optimized filling leads to the formation of contrasting artifacts. This can have a particularly negative effect when using glossy, silk, multicolored filaments, or when making lithographs or lithophanes.

<img width="1886" height="1079" alt="image" src="https://github.com/user-attachments/assets/dba21345-5245-4f33-a795-db191ff289ed" />

Some notes to ensure the cleanliness of the top surface:

- Set the filling density for the top and bottom surfaces to 75-90% to obtain a uniform layer without overflow on adjacent lines. Also, keep the patchwork_infill_flow_ratio value less than number one so that the tracks don't overlap when printing.
- Set more hotend temperature to add shine. Try, for example, set the 240°C Instead of 220°C recommended by the manufacturer for PLA plastic. Make sure that the model is printed without problems and unnecessary smudges.
- Use the brand new and quality nozzles.
- Set a low surface printing speed. Usually 20-30 mm/s. Choose according to the best result.
- Adjust the Pressure Advance
- Apply Z-hop at least to the top surface.
- Reduce the retracts. Too frequent pressure drops make melted filament mass less uniform.
- Don't use classic wipe method. That's may dirty top surface. If the extends of pathlines out the pattern beyond the contour of the model, then you can specify a negative value to wipe_distance parameter so that all the stretching snot threads freeze outside the model.
- Use dry plastic. Do not allow it to leak out of the nozzle when traveling.
- Increase the speed of movement and acceleration of the hotend during travel to the maximum possible. This will reduce the amount of snotty plastic. Make sure that the kinematics of the printer can handle such loads.
- Ensure that the top surface for PLA plastic is cooled as much as possible.

## Center of surface patterns

Previously, all centered patterns were aligned only individually on each surface, which did not ensure the integrity of the product in some artistic projects. Now you can select the infill alignment distribution:

### For each surface
The previous regime
<img width="1375" height="564" alt="image" src="https://github.com/user-attachments/assets/457f6b1d-8d57-40b6-b53a-1531e06f5c9d" />

### For each model
All the surfaces of one model (or each shape in the assembly) are combined
<img width="1375" height="564" alt="image" src="https://github.com/user-attachments/assets/1fba5a30-41e9-41a0-8843-cbe17fac77c5" />

### For each assembly
The mode where all the surfaces of the same assembly fall under the union. This property is well suited for articulated models.
<img width="1375" height="564" alt="image" src="https://github.com/user-attachments/assets/8f01f834-ce68-4a48-ab9a-dd93863a13d6" />

## Separated Infills
The same applies for the inner fill for the assembly. Previously, the entire assembly was considered as a single whole, and when applying a centered fill or rotation pattern, it resulted in it rotating around a common center.

### Separated Infills OFF
<img width="1375" height="564" alt="image" src="https://github.com/user-attachments/assets/a01e1c3c-96cb-4f3a-bdc4-642d0f1405a4" />

### Separated Infills ON
<img width="1375" height="564" alt="image" src="https://github.com/user-attachments/assets/199e7ce8-4853-4fd5-b3d2-eaea9171cad1" />

pi-squared-studio committed:
Pack of features

pi-squared-studio commented:
@valerii-bokhan OK, I did it right in PR. Look, everything should work.

RF47 committed:
Align formatting of comparison operators in Fill.cpp

RF47 commented on src/libslic3r/PrintConfig.cpp:
```diff
@@ -6720,6 +6725,38 @@ void PrintConfigDef::init_fff_params()
+                       "dispersion when using multi-colored or silk plastics. "
+                       "This option disable the gap fill. "
+                       "This option can increase a printing time. ");
+    def->mode     = comAdvanced;
```
```suggestion
    def->mode     = comExpert;
```

RF47 commented on src/libslic3r/PrintConfig.cpp:
```diff
@@ -6720,6 +6725,38 @@ void PrintConfigDef::init_fff_params()
+    def->label    = L("Separated infills");
+    def->category = L("Strength");
+    def->tooltip  = L("Separated infills for each part of the one model. ");
+    def->mode     = comAdvanced;
```
```suggestion
    def->mode     = comExpert;
```

RF47 commented on src/libslic3r/PrintConfig.cpp:
```diff
@@ -6720,6 +6725,38 @@ void PrintConfigDef::init_fff_params()
+    def->enum_labels.push_back(L("Each Surface"));
+    def->enum_labels.push_back(L("Each Model"));
+    def->enum_labels.push_back(L("Each Assembly"));
+    def->mode = comAdvanced;
```
```suggestion
    def->mode = comExpert;
```

ianalexis commented:
There are missing the new options in GUI_Factories.cpp

RF47 committed:
Expert mode

Co-authored-by: Rodrigo Faselli <162915171+RF47@users.noreply.github.com>

RF47 commented on src/libslic3r/PrintConfig.cpp:
```diff
@@ -6799,6 +6804,61 @@ void PrintConfigDef::init_fff_params()
+    def->sidetext = ("%");
+    def->min      = 10;
+    def->max      = 100;
+    def->set_default_value(new ConfigOptionPercent(100));
```
@pi-squared-studio , was this repeated?

RF47 commented:
Separated infills, simple assembly OK:
<img width="2560" height="1392" alt="image" src="https://github.com/user-attachments/assets/a0f7f067-e663-4a21-b2ee-952c0f4becff" />
~~separated infill (3 objects) doesnt work:~~ FIXED
<img width="2560" height="1392" alt="image" src="https://github.com/user-attachments/assets/1f17ceb5-d823-420d-b183-7e583ed7d2a8" />

separated infills (3 objects, not intersecting) OK:
<img width="2560" height="1392" alt="image" src="https://github.com/user-attachments/assets/e39fd75b-cd84-413a-806f-12aa04261f42" />

ianalexis committed:
Fixes

Co-Authored-By: Rodrigo Faselli <162915171+RF47@users.noreply.github.com>

ianalexis committed:
Fix empty parts infill

Co-Authored-By: Rodrigo Faselli <162915171+RF47@users.noreply.github.com>

ianalexis committed:
Separate infill centering for model groups

Improve infill centering logic when using per-model or separated infills. Replace previous per-volume selection with an overlap-area pick (best overlapping volume) and then compute the centering bbox from either the whole overlap group (for separated infill) or the single winning volume (for per-model centering). This handles assemblies where parts overlap or touch by using firstLayerObjGroups(), merges bounding boxes of group volumes, and preserves existing centering behavior when no group is present. Clarifies intent and simplifies instance/volume selection.

Co-Authored-By: Rodrigo Faselli <162915171+RF47@users.noreply.github.com>

ianalexis committed:
Fix Each Model overlaping objects in assembly

Clarify comments around per-model/separated-infill centering and simplify the selection logic for center volume IDs. The code comment now explains that centering is done per connected physical body (overlap group) for both separated infills and Each_Model centering. The conditional branch that only used groups for separated infills was removed so the group lookup is applied uniformly, falling back to the single winning part if no group is found. This is a documentation/clarity and minor-refactor change with no functional behavior change.

Co-Authored-By: Rodrigo Faselli <162915171+RF47@users.noreply.github.com>

ianalexis commented:
So using overlapping objects in the same assembly was causing some issues.
We fix it by using center of mass for that cases:

<img width="1098" height="1081" alt="imagen" src="https://github.com/user-attachments/assets/6ef6edce-ae3e-4c20-b196-c965f3598db6" />

Each Surface:
<img width="1266" height="1296" alt="imagen" src="https://github.com/user-attachments/assets/231a6792-41ca-4167-86cc-4a8350d0b523" />

Each Model:
<img width="1266" height="1296" alt="imagen" src="https://github.com/user-attachments/assets/98a627e1-1d14-4816-99ee-4f678609a47d" />

Each Assembly:
<img width="1266" height="1296" alt="imagen" src="https://github.com/user-attachments/assets/44290a6d-7d0b-4022-96ce-dff8114e9606" />

Same for Separated infills:

On:
<img width="1266" height="1296" alt="imagen" src="https://github.com/user-attachments/assets/378c3d04-4fe2-4ef0-9f2b-a962457b73b3" />

Off:
<img width="1266" height="1296" alt="imagen" src="https://github.com/user-attachments/assets/b1bc0dd2-c133-48fb-a38f-a3c33e96d8f6" />

Now we think this is ready @pi-squared-studio if you want to test it a little bit more you are welcome.

ianalexis committed:
Improve descriptions

ianalexis committed:
Wiki redirection

pi-squared-studio commented:
@ianalexis @RF47  Thank you for your work. I can't test it right now, as I want to improve the PR for Fuzzy skin (I'm trying to create a flawless drawing, so I had to convert the extrusion path to Clipper geometry to take advantage of vector graphics). There were a lot of changes that needed to be committed, and frequent Git switches were causing crashes. I hope to be able to test it soon.

ianalexis commented:
> @ianalexis @RF47 Thank you for your work. I can't test it right now, as I want to improve the PR for Fuzzy skin (I'm trying to create a flawless drawing, so I had to convert the extrusion path to Clipper geometry to take advantage of vector graphics). There were a lot of changes that needed to be committed, and frequent Git switches were causing crashes. I hope to be able to test it soon.

Dont worry! I really like this PR and think is necessary.
We tried to fix the issues we found keeping your logic.

ianalexis committed:
Toggle anisotropic line

 The co-directional reorientation only runs for the plane-path surface patterns (Archimedean Chords, Octagram Spiral, Hilbert Curve); with any other top/bottom pattern the option cannot make a surface anisotropic, so only expose it for those.

RF47 commented on src/slic3r/GUI/ConfigManipulation.cpp:
```diff
@@ -720,6 +720,14 @@ void ConfigManipulation::toggle_print_fff_options(DynamicPrintConfig *config, in
 
+    // Orca: anisotropic surfaces.
+    auto is_anisotropic_pattern = [](InfillPattern p) {
+        return p == InfillPattern::ipArchimedeanChords || p == InfillPattern::ipOctagramSpiral || p == InfillPattern::ipHilbertCurve;
```
ipHilbertCurve no va

ianalexis committed:
Simplify config manipulation
